### PR TITLE
window.print() support for WebEngine

### LINKED ISSUE #3495: window.print() support
We can probably hack in `window.print()` support similar to how QupZilla did it 😆 

https://github.com/QupZilla/qupzilla/commit/d3f0d766fb052dc504de2426d42f235d96b5eb60

### PR DESCRIPTION
**Pull request by sarehag**
An attempt to solve window.print() support in webengine. I hope i managed to follow the guidelines and way of coding appropriately. If something needs changing, then I would be happy to open a new pull request, if you give me some hints on what to do.

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3864)

*edit by @The-Compiler to link the issue: Closes #3495*

### CODE REVIEW

**Comment by The-Compiler** [comment]
Given that QupZilla/Falkon uses the same hack (https://github.com/QupZilla/qupzilla/commit/d3f0d766fb052dc504de2426d42f235d96b5eb60) and it has worked out fine for them in practice, I'm fine with the approach.

Might be a good idea to have a QtWebChannel in the future (though it'd need to be separate from a QtWebChannel in the non-main world; only used for things which are okay for the page to do), but I'd say we can do this in the meantime.

**Comment by The-Compiler** [comment]
I opened #3883 for looking at QtWebChannel in detail.

File: qutebrowser/browser/webengine/webenginetab.py [hunk 1]
```diff
@@ -700,6 +700,12 @@ def _init_js(self):
         self._inject_early_js('js', js_code, subframes=True)
         self._init_stylesheet()
 
+        js_code_print = utils.read_file('javascript/print.js')
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: No need for a separate variable here - just call `utils.read_file` directly below.

File: qutebrowser/browser/webengine/webview.py [hunk 2]
```diff
@@ -310,9 +311,17 @@ def acceptNavigationRequest(self,
             QWebEnginePage.NavigationTypeOther:
                 usertypes.NavigationRequest.Type.other,
         }
+        accept_request = True
+        if(url.scheme() == "qute" and url.path() == "print"):
+            command_dispatcher = objreg.get('command-dispatcher',
+                                            scope='window',
+                                            window=self._win_id)
+            command_dispatcher.printpage()
+            accept_request = False
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webview.py]: This is the wrong place to handle this (`webview.py` should be a pretty minimal wrapper around QWebEngineView). Instead, you should handle it in `webenginetab.py` (in `_on_navigation_request`) where you also already have the `win_id` available.

File: qutebrowser/browser/webengine/webview.py [hunk 3]
```diff
@@ -310,9 +311,17 @@ def acceptNavigationRequest(self,
             QWebEnginePage.NavigationTypeOther:
                 usertypes.NavigationRequest.Type.other,
         }
+        accept_request = True
+        if(url.scheme() == "qute" and url.path() == "print"):
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webview.py]: you should be able to do `if url == QUrl('qute://print')` instead.

File: qutebrowser/javascript/print.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+"use strict";
+
+window.print = function() {
+    window.location = "qute:print";
```
Comment (The-Compiler) [inline qutebrowser/javascript/print.js]: Use `qute://print` for consistency with other `qute://` URLs. Also, maybe add a comment about why we're doing this.

**Comment by sarehag** [comment]
Spoke with The-Compiler on IRC and he said that no changes should need to be done to webengine/webview.py, and it should be sufficient to make changes in _on_navigation_request in webengine/webenginetab.py.

I have been unable to find a way to stop the navigation here. It seems to have happened already and seemingly whatever i do i get a ERR_FILE_NOT_FOUND error. I also tried in the interceptor, but i get another error,ERR_BLOCKED_BY_CLIENT, displayed in the browser if i use info.block(True). Url is also changed in both cases.

I also read this http://doc.qt.io/qt-5/qwebenginepage.html#acceptNavigationRequest, which I interpret as if the browser will load the url unless the acceptNavigationRequest function in webview returns false.

I am sorry for consuming so much of your time with an issue tagged easy, but I cannot seem to find a way to achieve acceptable behavior when only modifying webenginetab.py. I would humbly accept any hints or pointers you may have.

**Comment by The-Compiler** [comment]
No problem! :smile: 

In `_on_navigation_request` you should be able to simply do `navigation.accepted = False` - then the code in `webview.py` should take care of aborting the request.

**Comment by sarehag** [comment]
As for adding a test, I started looking at the misc.feature tests for printing, but the ones checking for the opening of a print dialogue, which makes sense to check, seem to be disabled, (@xfail_norun). 

I started with adding a html-file under end2end/data that runs window.print from javascript, and I think I could make a bdd-test open the file, this seem to be done a lot in other tests. As for the remainder of the test, I may need a pointer or so, since the other tests for print functionality are disabled.

**Comment by The-Compiler** [comment]
Thanks for the PR and your patience - looks good now! :+1: 

I don't think it's possible to test this in an automated fashion - the print dialog is quite cumbersome to handle in tests, as it can be a native OS dialog, so an automated test can't easily close it automatically after it's opened.

**Pull request merged**